File: matlab/examples_paper/algorithms/AML.m
function [out] = AML(R_tilde, N, Psi)
%
% INPUT: 
%         R_tilde: sample covariance of dimension M.
%               N: Number of data points used to build the sample covariance.
% OUTPUT:
%         out.x: real part first row of Toeplitz matrix
%         out.y: imaginary part first row of Toeplitz matrix
%         out.T: Toeplitz matrix
%
% Note:
% *  If out.x = (x0, x1, x2, ..., xm), out.y = (y1, y2, ..., ym), then
%    out.T is the Hermitian Toeplitz matrix with first row equal to
%    (x0, x1+1i*y1, x2+1i*y2, ..., xm+1i*ym).
% out.T = [x0   x1+1i*y1 
% * This estimator assumes that N >= m.

tic;
M = size(R_tilde, 1);
r_tilde = R_tilde(:);
C_tilde =  1/N*kron(transpose(R_tilde), R_tilde);
C_tilde_chol = chol(C_tilde, 'lower');
%phi_tilde = (Psi'*(C_tilde\Psi))\(Psi'*(C_tilde\r_tilde));
phi_tilde = (Psi'*(C_tilde_chol'\(C_tilde_chol\Psi)))\(Psi'*(C_tilde_chol'\(C_tilde_chol\r_tilde)));
phi_tilde = real(phi_tilde); % Remove negligible imaginary parts.
x = [phi_tilde(1); phi_tilde(2:2:2*M-2)];
y = phi_tilde(3:2:2*M-1);
T = toeplitz(x + 1i*[0;y]);
out.x = x;
out.y = y;
out.estimate = T;
out.solve_time = toc;
end

File: python/examples_paper/algorithms/aml.py
"""Asymptotic Maximum Likelihood (AML) Toeplitz estimator."""
import time
import numpy as np
from scipy.linalg import toeplitz


def create_psi_aml(M):
    """Build the AML constraint matrix Psi.

    Parameters
    ----------
    M : int
        Dimension of the covariance matrix.

    Returns
    -------
    Psi : ndarray, shape (M*M, 2*M-1)
    """
    K = 2 * M - 1
    Omega = np.zeros((K, K), dtype=complex)
    Omega[0, 0] = 1
    col_counter = 1
    for row in range(1, K):
        if row % 2 == 1:  # even row in 1-based (row+1 is even)
            Omega[row, col_counter] = 1
            Omega[row, col_counter + 1] = 1j
        else:  # odd row in 1-based
            Omega[row, col_counter] = 1
            Omega[row, col_counter + 1] = -1j
            col_counter += 2

    Sigma = np.zeros((M * M, K), dtype=complex)
    IM = np.eye(M)
    Sigma[:, 0] = IM.flatten("F")
    col_counter = 1
    for m in range(1, M):
        Q_col = np.zeros((M, M))
        Q_col[:M - m, m:] = np.eye(M - m)
        Sigma[:, col_counter] = Q_col.flatten("F")
        Sigma[:, col_counter + 1] = Q_col.T.flatten("F")
        col_counter += 2

    return Sigma @ Omega


def aml(R_tilde, N, Psi):
    """Asymptotic Maximum Likelihood Toeplitz estimator.

    Parameters
    ----------
    R_tilde : ndarray, shape (M, M)
        Sample covariance matrix.
    N : int
        Number of data samples.
    Psi : ndarray, shape (M*M, 2*M-1)
        Constraint matrix from create_psi_aml.

    Returns
    -------
    dict with keys 'estimate', 'x', 'y', 'solve_time'.
    """
    t0 = time.perf_counter()
    M = R_tilde.shape[0]
    r_tilde = R_tilde.flatten("F")
    C_tilde = (1 / N) * np.kron(R_tilde.T, R_tilde)
    L = np.linalg.cholesky(C_tilde)
    Linv_Psi = np.linalg.solve(L, Psi)
    Linv_r = np.linalg.solve(L, r_tilde)
    phi_tilde = np.linalg.solve(Linv_Psi.conj().T @ Linv_Psi,
                                Linv_Psi.conj().T @ Linv_r)
    phi_tilde = np.real(phi_tilde)
    x = np.concatenate([[phi_tilde[0]], phi_tilde[1::2][:M - 1]])
    y = phi_tilde[2::2][:M - 1]
    T = toeplitz(x + 1j * np.concatenate([[0], y]))
    solve_time = time.perf_counter() - t0
    return {"estimate": T, "x": x, "y": y, "solve_time": solve_time}
